The code used for data analysis in the paper 'Aberrant regulation of alternative pre-mRNA splicing in Rett syndrome' is available at https://github.com/HJY-Lab/MeCP2_splicing.

The workflows of RNA-Seq and ChIP-Seq are based on Snakemake (https://snakemake.readthedocs.io/en/stable/).

The workflow of alternative splicing is based on rMATS (http://rnaseq-mats.sourceforge.net/). It needs BAM files of the aligned RNA-Seq reads which are generated by RNA-Seq workflow.

The iCLIP workflow is based on the pipeline developed by Busch, A. et al (https://www.sciencedirect.com/science/article/pii/S1046202318304948).
